### DISCUSSION TITLE
Implementing the divergence with deepxde

### DISCUSSION THREAD

**Discussion by AudreyDemafo** [lululxvi/deepxde#1301] [General]
Hello everyone,

I am to deepxde. I am trying to solve the continuity equation for an incompressible fluid in 2D. But I am getting an error that I am yet to fix

Here is the equation:
du_dx + dv_dy -2 = 0

My code attempt is the following:

import matplotlib.pyplot as plt
from matplotlib import cm
import numpy as np

import deepxde as dde

def pde(u, v, x):       # Continuity equation in fluid mechanics
  du_dx = dde.gradients.jacobian(u, x, i=0, j=0)
  dv_dy = dde.gradients.jacobian(v, x, i=0, j=1)
  return du_dx - dv_dy - 2

# Define boundaries of the domain
def lower_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[1], 0)

def upper_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[1], 2)

# Define the geometry and apply the boundary conditions
geom = dde.geometry.Rectangle(xmin = [0,0], xmax=[1,2]) # [0,0] :--> bottom left corner while [1,2] :--> Top right corner of the rectangle 

# Apply the boundary equations
bc1 = dde.DirichletBC(geom, lambda x: 5, left_boundary)
bc2 = dde.DirichletBC(geom, lambda x: 0, lower_boundary)
bc3 = dde.DirichletBC(geom, lambda x: 5, right_boundary)
bc4 = dde.DirichletBC(geom, lambda x: 0.75, upper_boundary)

# Data and Neural network architecture
data = dde.data.PDE(
    geom,
    pde,
    [bc1, bc2, bc3, bc4],
    num_domain = 500,
    num_boundary = 250,

)

nnet = dde.maps.FNN([2] + [100]*5 + [1], "tanh", "Glorot uniform") 
model = dde.Model(data, nnet)

model.compile("adam", lr = 0.0001, loss="mse") 

model.train(epochs=20000)

losshistory, train_state = model.train(epochs=20000)

dde.saveplot(
    losshistory, train_state, issave=True, isplot=True, output_dir=OUTPUT_DIRECTORY
)

def left_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[0], 0)

def right_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[0], 1)

and the error is:
--> ValueError                                Traceback (most recent call last)
[<ipython-input-9-9a0d6a6e18b8>](https://localhost:8080/#) in <cell line: 14>()
     12 model = dde.Model(data, nnet)
     13 
---> 14 model.compile("adam", lr = 0.0001, loss="mse")
     15 
     16 model.train(epochs=20000)

5 frames
[/usr/local/lib/python3.10/dist-packages/deepxde/data/pde.py](https://localhost:8080/#) in losses(self, targets, outputs, loss_fn, inputs, model, aux)
    129                 if self.auxiliary_var_fn is None:
    130                     if aux is None or len(aux) == 1:
--> 131                         raise ValueError("Auxiliary variable function not defined.")
    132                     f = self.pde(inputs, outputs_pde, unknowns=aux[1])
    133                 else:

-->ValueError: Auxiliary variable function not defined.

Any help on this will be very much appreciated. Thank you in advance

**Comment by pescap** [comment]
Hi, to facilitate reading can you format your code using [code blocks ```](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)? 

```
your
code 
```

**Reply by AudreyDemafo** [reply]
```

import matplotlib.pyplot as plt
from matplotlib import cm
import numpy as np

import deepxde as dde

def pde(y, x):         # Continuity equation in fluid mechanics
  #u = y[:, 0:1]
  #v = y[:, 1:2]       
  du_dx = dde.grad.jacobian(y, x, i=0, j=0)
  dv_dy = dde.grad.jacobian(y, x, i=1, j=1)                                  
  return du_dx + dv_dy - 2

# Define boundaries of the domain
def lower_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[1], 0)

def upper_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[1], 2)

def left_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[0], 0)

def right_boundary(x, on_boundary):
  return on_boundary and np.isclose(x[0], 1)

# Define the geometry and apply the boundary conditions
geom = dde.geometry.Rectangle(xmin = [0,0], xmax=[1,2]) # [0,0] :--> bottom left corner while [1,2] :--> Top right corner of the rectangle 

# Apply the boundary equations
bc1 = dde.DirichletBC(geom, lambda x: 5, left_boundary, component=0)
bc2 = dde.DirichletBC(geom, lambda x: 0, lower_boundary, component=1)
bc3 = dde.DirichletBC(geom, lambda x: 5, right_boundary, component=0)
bc4 = dde.DirichletBC(geom, lambda x: 0.75, upper_boundary, component=1)

# Data and Neural network architecture
data = dde.data.PDE(
    geom,
    pde,
    [bc1, bc2, bc3, bc4],
    num_domain = 500,
    num_boundary = 250,

)

nnet = dde.maps.FNN([2] + [100]*5 + [1], "tanh", "Glorot uniform") 
model = dde.Model(data, nnet)

model.compile("adam", lr = 0.0001, loss="mse") 

model.train(epochs=200, display_every=20)

losshistory, train_state = model.train(epochs=20000)

dde.saveplot(
    losshistory, train_state, issave=True, isplot=True, output_dir=OUTPUT_DIRECTORY
)

```

Hi @pescap, thank you for your reply. Here is the code formatted.

**Comment by AudreyDemafo** [comment]
The problem was the number of layer units in the output layer. I was taking 1 instead of 2 since I wanted the code to output the two velocity components.

Thank you.
